### ISSUE TITLE
Evaluation Metrics

### ISSUE THREAD

**Issue by alif2499** [FareedKhan-dev/train-llm-from-scratch#10]
Can you provide the code to evaluate the model performance on different evaluation metrics?

**Comment by ZubairAli999** [comment]
```
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    confusion_matrix,
    classification_report
)

import numpy as np


# -----------------------------
# 1. Evaluation Function
# -----------------------------
def evaluate_model(model, X_test, y_test, average="weighted"):

    y_pred = model.predict(X_test)

    metrics = {
        "accuracy": accuracy_score(y_test, y_pred),
        "precision": precision_score(y_test, y_pred, average=average),
        "recall": recall_score(y_test, y_pred, average=average),
        "f1_score": f1_score(y_test, y_pred, average=average),
    }

    # ROC-AUC (if supported)
    if hasattr(model, "predict_proba"):
        y_proba = model.predict_proba(X_test)

        try:
            if len(np.unique(y_test)) == 2:
                metrics["roc_auc"] = roc_auc_score(y_test, y_proba[:, 1])
            else:
                metrics["roc_auc"] = roc_auc_score(
                    y_test, y_proba, multi_class="ovr", average=average
                )
        except:
            metrics["roc_auc"] = None

    metrics["confusion_matrix"] = confusion_matrix(y_test, y_pred)

    return metrics


# -----------------------------
# 2. Pretty Print Function
# -----------------------------
def print_metrics(metrics):

    print("\n========== MODEL EVALUATION REPORT ==========\n")

    for key, value in metrics.items():
        if key != "confusion_matrix":
            print(f"{key.upper():<15}: {value}")

    print("\nCONFUSION MATRIX:")
    print(metrics["confusion_matrix"])


# -----------------------------
# 3. Load Dataset
# -----------------------------
data = load_breast_cancer()
X = data.data
y = data.target

# Split dataset
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# -----------------------------
# 4. Train Model
# -----------------------------
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# -----------------------------
# 5. Evaluate Model
# -----------------------------
metrics = evaluate_model(model, X_test, y_test)

# -----------------------------
# 6. Print Results
# -----------------------------
print_metrics(metrics)
```

**Comment by FareedKhan-dev** [comment]
Thanks for raising this! Model evaluation is now built into the repo, so I'll close this as resolved.

The headline metric across the pipeline is **GSM8K accuracy (greedy decoding)**, so you can compare Base → SFT → DPO → PPO → GRPO on one axis:

- `scripts/eval_post_training.py` — evaluate any checkpoint on GSM8K and print a stage-comparison table:
  ```bash
  PYTHONPATH=. python scripts/eval_post_training.py --ckpt /path/to/model.pt --label sft --limit 200
  ```
- `src/post_training/evaluation.py` — reusable helpers: `gsm8k_accuracy`, `batched_generate`, `load_gsm8k_eval`.
- The Streamlit **Evaluate** page (`ui/pages/8_Evaluate.py`) — run GSM8K eval and inspect sample generations.
- For base pretraining, `estimate_loss()` in `scripts/train_transformer.py` reports train/dev loss during training (perplexity follows directly).
- See `docs/08_evaluation.md` for the methodology.

(Note: the `sklearn` accuracy/precision/recall/F1 snippet posted above is for tabular classification, not a generative LM, so it isn't applicable here.)

Closing as resolved — happy to reopen if you want a specific additional metric (e.g. standalone perplexity reporting or another benchmark). 🙏
